File: share/example/c++/example_script.cpp
#include "aubo_sdk/rpc.h"
#include "aubo_sdk/script.h"
#include <fstream>
#include <thread>
#ifdef WIN32
#include <Windows.h>
#endif
using namespace arcs::common_interface;
using namespace arcs::aubo_sdk;
using namespace std;
/**
 * Function: Run local script program
 * Steps:
 * Step 1: Connect to RPC service and log in to the robot arm
 * Step 2: Connect to SCRIPT service and log in to the robot arm
 * Step 3: Enter the script file name or absolute path
 * Case 1: If the lua script is copied to the executable program path (build/bin), enter the script file name
 * Case 2: If the lua script is not in the executable program path (build/bin), enter the absolute path of the script
 * Step 4: Read the file and run the script program. If opening the file fails, exit the program.
 */

#define LOCAL_IP "127.0.0.1"

int main(int argc, char **argv) {
#ifdef WIN32
  // Set Windows console output code page to UTF-8
  SetConsoleOutputCP(CP_UTF8);
#endif
  auto rpc = std::make_shared<RpcClient>();
  // API call: Set RPC timeout
  rpc->setRequestTimeout(1000);
  // API call: Connect to RPC service
  rpc->connect(LOCAL_IP, 30004);
  // API call: Login
  rpc->login("aubo", "123456");

  auto script = std::make_shared<ScriptClient>();
  // API call: Connect to SCRIPT service
  script->connect(LOCAL_IP, 30002);
  // API call: Login
  script->login("aubo", "123456");

  // Enter script file name
  char file_name[20];
  cin >> file_name;

  // Open file
  ifstream file;
  file.open(file_name);
  // If opening the file fails, exit the program
  if (!file) {
    cout << "open fail." << endl;
    exit(1);
  }

  // Read and print the contents of the script file
  std::string str_all;
  while (!file.eof()) {
    std::string str_line;
    getline(file, str_line);
    str_all += str_line;
    str_all += "\n";
  }

  // Find "\r\n\r\n" in the string
  size_t found = str_all.find("\r\n\r\n");

  // Loop to find and replace all matches
  while (found != std::string::npos) {
    str_all.replace(found, 4, "\n"); // 4 is the length of "\r\n\r\n"
    found = str_all.find("\r\n\r\n", found + 1);
  }
  str_all += "\r\n\r\n";
  cout << "Script content:" << endl << str_all << endl;
  file.close();

  // API call: Send local script content to controller
  script->sendString(str_all);

  // Wait for the planner to start
  int i = 0;
  while (1) {
    if (i++ > 5) {
      std::cerr << "Planner failed to start" << std::endl;
      return -1;
    }
    if (rpc->getRuntimeMachine()->getStatus() == RuntimeState::Running) {
      std::cout << "Planner started successfully, start executing script" << std::endl;
      break;
    }
    std::this_thread::sleep_for(std::chrono::milliseconds(5));
  }

  // Wait for the planner to stop
  while (1) {
    if (rpc->getRuntimeMachine()->getStatus() == RuntimeState::Stopped) {
      std::cout << "Planner stopped, script execution completed" << std::endl;
      break;
    }
    std::this_thread::sleep_for(std::chrono::milliseconds(5));
  }

  // API call: Logout
  rpc->logout();
  // API call: Disconnect
  rpc->disconnect();

  // API call: Logout
  script->logout();
  // API call: Disconnect
  script->disconnect();

  return 0;
}

File: share/example/python/example_script.py
#! /usr/bin/env python
# coding=utf-8

"""
Run script

Steps:
Step 1: Connect to RPC service, robot login
Step 2: Connect to SCRIPT service, robot login
Step 3: Run script
Step 4: When the planner stops running, stop script execution
"""
import time

import pyaubo_sdk

robot_ip = "127.0.0.1"  # Server IP address
M_PI = 3.14159265358979323846
robot_rpc_client = pyaubo_sdk.RpcClient()
robot_script_client = pyaubo_sdk.ScriptClient()


def exampleScript():
    robot_script_client.sendFile("../aubo_script/example_movej.lua")
    pass


if __name__ == '__main__':
    robot_rpc_client.connect(robot_ip, 30004)  # API call: connect to RPC service
    robot_rpc_client.login("aubo", "123456")  # API call: robot login
    robot_script_client.connect(robot_ip,30002)  # API call: connect to SCRIPT
    robot_script_client.login("aubo", "123456")  # API call: robot login

    exampleScript()  # Run script

    while True:
        time.sleep(1)
        runtime_state = robot_rpc_client.getRuntimeMachine().getStatus()
        if runtime_state != pyaubo_sdk.RuntimeState.Running:  # When the planner stops running, stop script execution
            print("Planner state:", runtime_state)
            break
        print("Planner state:", runtime_state)